library/core/src/result.rs

```diff
@@ -853,6 +853,35 @@ impl<T, E> Result<T, E> {
             Err(e) => Err(op(e)),
         }
     }
+    
+    /// Returns `Err(e)` if the result is `Err(e)` (where `e` is the wrapped error),
+    /// otherwise calls `predicate` with the wrapped value and returns:
+    ///
+    /// - `Ok(t)` if `predicate` returns `Ok(())` (where `t` is the wrapped value), and
+    /// - `Err(e)` if `predicate` returns `Err(e)`.
+    ///
+    /// This function is analogous to [`Option::filter`].
```

Comment:
It really is - the motivation for this was `Option::filter`. The whole doc is basically lifted from `Option::filter`. When we say it's analogous, we mean it.

Reply:
"inspired by" may be correct, but is not the same as "analogous", which suggests a stronger similarity. It is important not to claim excess similarity because it can confuse a user who does not see things exactly the same way. Moreover, this function cannot help but be quite different because it is allowed to reinterpret an Ok as an Err, but Err requires a paired inner value to be meaningful. It does not merely eliminate possible values, as with filter, it can add to them. In other words:

<details>
<summary><code>Option::filter</code> is monotonically decreasing.</summary>
Option includes Some and None
Some has an inner value x
filter evaluates a function on Option
filter sends Some x to Some v or None
fllter sends None to None
filter may return Some v or None
v is a subset of x
thus, filter has reduced* the possible set of values for an Option.

\* "reduced by zero" may be true if the predicate is always true!
</details>

<details>
<summary><code>Result::check</code> is non-monotonic.</summary>
Result includes Ok and Err
Ok has an inner value x
Err has an inner value y
check evaluates a function on Result
check sends Ok x to Ok v or Err z
check sends Err y to Err y
check may return **Ok v, Err y, or Err z**
v is a subset of x
thus, check has reduced the possible domain of Ok values...
**but** z has no defined relationship with y: they merely must share a type
thus, the possible set of Err values may grow or shrink
thus, check may grow or reduce the possible set of values for a Result.
</details>

Consider this example:
```rust
#![feature(result_check)]
use rand::Rng;

fn neg_or_rand(x: &i32) -> Result<(), i32>{
    let mut rng = rand::thread_rng();
    (x < &0).then(|| ()).ok_or_else(|| rng.gen())
}

let a = Ok(0);
let b = Ok(-1);
let c = Result::check(a, neg_or_rand);
let d = Result::check(b, neg_or_rand);
assert_ne!(a, c);
assert_eq!(b, d);
```

If we had any knowledge about the interior values of Err, here, we now know much less after using `check` here. This is to be expected, because Result is a type some languages call Either, with less implied attachment to it meaning a "possibly erroneous" value. While there is no reason to adopt that convention for Rust and suddenly rename the type, it is useful to keep in mind sometimes at the purely logical level... notably, it can affect API design and how we talk about them.

Reply:
We still maintain that it's designed/intended to be analogous to `Option::filter`. Specifically, it's meant to be the most straightforward adaptation of the `Option::filter` semantics onto `Result`. The only difference is that `Result` requires providing an error when you discard the value.